### PR TITLE
[Gemma Quant] Preserve Gemma 4 MoE router projection in full precision during quantization

### PR DESCRIPTION
**Pull request by pculliton**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48801) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48801)

# What does this PR do?

In Gemma 4 MoE models (e.g. Gemma 4 26B), the router projection `router.proj` is an `nn.Linear` layer. During quantization with `replace_with_quant_layers` or `GemmaQuantizer`, `router.proj` was being matched by generic linear quantization rules and converted to `QuantizedLinear` (INT4/INT2).

Routing logits are sensitive to precision degradation:
- Quantizing `router.proj` to INT4 causes top-$k$ expert selection agreement to collapse to **86.1%** (~1 in 7 tokens misrouted).
- Quantizing `router.proj` to INT2 causes top-$k$ agreement to collapse to **56.8%** (~half of tokens misrouted).
- The router projection across all 30 layers accounts for only **~21.6 MB** (<0.04% of model weights), providing practically zero memory benefit when quantized.

This PR ensures that `router.proj` remains in full precision (`bfloat16`/`float32`):
1. Adds `_keep_in_fp32_modules = ["router.proj"]` to `Gemma4PreTrainedModel` in `modular_gemma4.py` and `modeling_gemma4.py`.
2. Updates `GemmaQuantizer` to safely pass `model._keep_in_fp32_modules` to `get_modules_to_not_convert`.
3. Updates `replace_with_quant_layers` to automatically preserve router projections in full precision by default, preventing unintended quantization even when called directly without an explicit `modules_to_not_convert` list.
4. Adds unit tests in `tests/quantization/gemma_integration/test_gemma.py`.

cc @SunMarc @vasqu @sgerrard

### CODE REVIEW

**Review by vasqu — COMMENTED** [review]
Hmm I mean yes it can make sense but you can also control it for yourself setting modules to not convert no?

Otherwise you fully restrict something that might be intended on the google side

**Comment by pculliton** [comment]
> Hmm I mean yes it can make sense but you can also control it for yourself setting modules to not convert no?
> 
> Otherwise you fully restrict something that might be intended on the google side

I am on the Google side - I'm the person who wrote the HfQuantizer implementation for Gemma 4. (Hello! :-) ) I'd like to fix the routing quant to resolve some quality issues we've seen with the quantized MoE model.

File: src/transformers/models/gemma4/modular_gemma4.py [hunk 1]
```diff
@@ -1261,6 +1261,7 @@ class Gemma4PreTrainedModel(Gemma3nPreTrainedModel):
         "Gemma4VisionPatchEmbedder",
         "Gemma4AudioLayer",
     ]
+    _keep_in_fp32_modules = ["router.proj"]
```
Comment (vasqu) [inline src/transformers/models/gemma4/modular_gemma4.py:1264]: I think you want it to be in the strict variation? This one is for fp16 only

Just as a thought

**Review by vasqu — COMMENTED** [review]
Oops sorry about that! :D hello

I'd still appreciate if we could rather fix it on the remote hub side to add on top of `modules_to_not_convert` (which is the purpose of this entry) rather than to manipulate behavior from the get go, e.g. [here](https://huggingface.co/google/gemma-4-E2B-it-qat-mobile-transformers/blob/main/config.json#L91)

File: src/transformers/models/gemma4/modular_gemma4.py (hunk 1 above)
Comment (pculliton) [reply src/transformers/models/gemma4/modular_gemma4.py:1264]: Thanks! I'll take a look.... would be a much simpler PR I think? :) Sorry, forgot a lot of the HfQuantizer context in the intervening months.

I'd actually love to grab some time with you at some point - given my previous PR (the Gemma HfQuantizer kernels) it'd be great to be more plugged into what you all are working on on your side. I'm figuring out OSS quants for ongoing Gemma models and shared context would be awesome!

Comment (vasqu) [reply src/transformers/models/gemma4/modular_gemma4.py:1264]: No worries :) take your time, things move fast in the field

If you do want the triton kernels faster it might make more sense to open a PR in kernels-community (it's the way we ship kernels) and we can load it in transformers then (similar to e.g. fp8 integrations)
- https://huggingface.co/docs/kernels/builder/writing-kernels - how it's meant to be done (locally)
- https://github.com/huggingface/kernels-community - where we maintain kernels
- Ideally we can also move the kernels under google then

For what we try to do is a bit ambitious but its split into
1. https://github.com/huggingface/kernels-community/pull/1018#top - the actual kernels implementation
2. https://github.com/huggingface/transformers/pull/48058 - the transformers side quantization and kernel wrapping

Our main goal is to support most quants in those kernels as the application mostly differs on the load / unpack side of things but the core across optimizations is very similar

Comment (vasqu) [reply src/transformers/models/gemma4/modular_gemma4.py:1264]: Btw are you on the shared slack channel? Maybe we can discuss more easily there?

Comment (vasqu) [reply src/transformers/models/gemma4/modular_gemma4.py:1264]: also cc @IlyasMoutawwakil

Comment (pculliton) [reply src/transformers/models/gemma4/modular_gemma4.py:1264]: Trying to get back into the slack :-) I've got an account but it's been a while. I'll ping as soon as I'm back in.
